modules/tftp/server.rb

```diff
@@ -90,21 +103,114 @@ def pxe_default
     end
   end
 
   class Pxegrub2 < Server
-    def pxeconfig_dir
-      "#{path}/grub2"
+    def bootloader_path(os, release, arch)
+      [release, "default"].each do |version|
+        bootloader_path = File.join(path, 'bootloader-universe/pxegrub2', os, version, arch)
+
+        logger.debug "TFTP: Checking if bootloader universe is configured for OS '#{os}' version '#{version}' (#{arch})."
+        logger.debug "TFTP: Checking if directory '#{bootloader_path}' exists."
+
+        return bootloader_path if Dir.exist?(bootloader_path)
+
+        logger.debug "TFTP: Directory '#{bootloader_path}' does not exist."
+      end
+      nil
+    end
+
+    def create_bootloader_universe_symlinks(bootloader_path, pxeconfig_dir_mac)
+      symlinks = []
+      Dir.glob(File.join(bootloader_path, '*.efi')).each do |source_file|
+        symlinks << { source: Pathname.new(source_file), symlink: Pathname.new(File.join(pxeconfig_dir_mac, File.basename(source_file))) }
+      end
+      create_symlinks(symlinks)
+    end
+
+    def create_default_symlinks(bootfile_suffix, pxeconfig_dir_mac)
+      pxeconfig_dir = pxeconfig_dir()
+
+      grub_source = "grub#{bootfile_suffix}.efi"
+      shim_source = "shim#{bootfile_suffix}.efi"
+
+      symlinks = [
+        { source: grub_source, symlink: "boot.efi" },
+        { source: grub_source, symlink: grub_source },
+        { source: shim_source, symlink: "boot-sb.efi" },
+        { source: shim_source, symlink: shim_source },
+      ]
+      symlinks.each do |link|
+        link[:source] = Pathname.new(File.join(pxeconfig_dir, link[:source]))
+        link[:symlink] = Pathname.new(File.join(pxeconfig_dir_mac, link[:symlink]))
+      end
+
+      create_symlinks(symlinks)
+    end
+
+    def create_symlinks(symlinks)
+      symlinks.each do |link|
+        relative_source_path = link[:source].relative_path_from(link[:symlink].parent)
+
+        logger.debug "TFTP: Creating relative symlink: #{link[:symlink]} -> #{relative_source_path}"
+        FileUtils.ln_s(relative_source_path.to_s, link[:symlink].to_s, force: true)
+      end
+    end
+
+    # Configures bootloader files for a host in its host-config directory
+    #
+    # @param mac [String] The MAC address of the host
+    # @param os [String] The lowercase name of the operating system of the host
+    # @param release [String] The major and minor version of the operating system of the host
+    # @param arch [String] The architecture of the operating system of the host
+    # @param bootfile_suffix [String] The architecture specific boot filename suffix
+    def setup_bootloader(mac:, os:, release:, arch:, bootfile_suffix:)
+      pxeconfig_dir_mac = pxeconfig_dir(mac)
+
+      logger.debug "TFTP: Deploying host specific bootloader files to '#{pxeconfig_dir_mac}'."
+
+      FileUtils.mkdir_p(pxeconfig_dir_mac)
+      FileUtils.rm_f(Dir.glob("#{pxeconfig_dir_mac}/*.efi"))
```

Comment:
I added this to guarantee that we have a clean directory state on each rebuild of a host in case the files required in the host_config directory change in future.

Reply:
This only applies if the OS changes, as the bootloader names remain the same if the OS is unchanged, so there's no issue in that case. My goal is to minimize the risk of accidentally deleting necessary files (e.g., the binaries in `/var/lib/tftpboot/grub2`). However, since we validate the mac address beforehand (see [here](https://github.com/theforeman/smart-proxy/blob/develop/modules/tftp/tftp_api.rb#L16)), I believe we’re safe in this regard.

Reply:
I took another and again deeper look at what happens regarding MAC address validation when the `setup_bootloader` method is called and yes: The way `setup_bootloader` is called guarantees that the `FileUtils.rm_f` command can and will only delete `*.efi` files in the `:tftproot:/host-config/<MAC_address>/` directory.
- Calls to the [post method of the Proxy::Api class](https://github.com/theforeman/smart-proxy/pull/877/files#diff-35f6ced397569968379e52db481cad28f591c91a6369e12730b854fe716d4098R51-R53) for `VARIANTS` others than Syslinux require a MAC address.
- And the MAC address will be validated in the `instantiate` method in the [line mentioned by you in your comment](https://github.com/theforeman/smart-proxy/pull/877/files#diff-35f6ced397569968379e52db481cad28f591c91a6369e12730b854fe716d4098R16).
- And last but not least: The [setup_bootloader method of the PxeGrub2 Server class](https://github.com/theforeman/smart-proxy/pull/877/files#diff-db3f0d8d32c77f2719de93fd56902295e1d21a1a1c0dc1e19f398cf94d69426eR165) requires the MAC address to be set, which will guarantee that the `pxeconfig_dir_mac` does not point to the `:tftproot:/grub2/` directory. This would be the case if the MAC address is `nil` when calling the [pxeconfig_dir method](https://github.com/theforeman/smart-proxy/pull/877/files#diff-db3f0d8d32c77f2719de93fd56902295e1d21a1a1c0dc1e19f398cf94d69426eR190-R195).

Although I have to admit that the cases I am thinking about here might be really rare corner cases [1], I would prefer to stick to deleting the present `*.efi` files on updating the symlinks.

[1] An example would be when an OS vendor decides to follow another naming scheme for the shim and GRUB2 binaries.